From f5f3cd970b7327ec16fefd17ac3b3606dfae5fe4 Mon Sep 17 00:00:00 2001
From: Ethan Chen <intervigil@gmail.com>
Date: Sun, 19 Nov 2017 15:38:24 -0800
Subject: [PATCH 14/44] fs_config: Allow targets to override OEM AID
 restrictions

* Some legacy devices set OEM AIDs that fall outside the Android O OEM
  AID ranges. Since legacy devices cannot necessarily change AIDs for
  their binaries easily, allow bypassing the AID restrictions if needed.

Change-Id: I13a7bc2036272b43f9424708aaacd894c8b1a65e
Signed-off-by: Kshitij Gupta <kshitijgm@gmail.com>
---
 tools/fs_config/Android.mk             | 12 +++++++---
 tools/fs_config/fs_config_generator.py | 40 +++++++++++++++++++++++++++++-----
 2 files changed, 44 insertions(+), 8 deletions(-)

diff --git a/tools/fs_config/Android.mk b/tools/fs_config/Android.mk
index 0e0b1da..8990ed8 100644
--- a/tools/fs_config/Android.mk
+++ b/tools/fs_config/Android.mk
@@ -48,6 +48,12 @@ LOCAL_REQUIRED_MODULES := \
 include $(BUILD_PHONY_PACKAGE)
 
 
+ifneq ($(TARGET_FS_CONFIG_GEN),)
+ifeq ($(TARGET_ALLOW_LEGACY_AIDS),true)
+allow_legacy_aids := --allow-legacy-aids
+endif
+endif
+
 ##################################
 # Generate the <p>/etc/fs_config_files binary files for each partition.
 # Add fs_config_files to PRODUCT_PACKAGES in the device make file to enable.
@@ -397,7 +403,7 @@ oem := $(local-generated-sources-dir)/generated_oem_aid.h
 $(oem): PRIVATE_LOCAL_PATH := $(LOCAL_PATH)
 $(oem): PRIVATE_TARGET_FS_CONFIG_GEN := $(TARGET_FS_CONFIG_GEN)
 $(oem): PRIVATE_ANDROID_FS_HDR := $(system_android_filesystem_config)
-$(oem): PRIVATE_CUSTOM_TOOL = $(PRIVATE_LOCAL_PATH)/fs_config_generator.py oemaid --aid-header=$(PRIVATE_ANDROID_FS_HDR) $(PRIVATE_TARGET_FS_CONFIG_GEN) > $@
+$(oem): PRIVATE_CUSTOM_TOOL = $(PRIVATE_LOCAL_PATH)/fs_config_generator.py oemaid --aid-header=$(PRIVATE_ANDROID_FS_HDR) $(allow_legacy_aids) $(PRIVATE_TARGET_FS_CONFIG_GEN) > $@
 $(oem): $(TARGET_FS_CONFIG_GEN) $(LOCAL_PATH)/fs_config_generator.py
 	$(transform-generated-source)
 
@@ -423,7 +429,7 @@ $(LOCAL_BUILT_MODULE): PRIVATE_TARGET_FS_CONFIG_GEN := $(TARGET_FS_CONFIG_GEN)
 $(LOCAL_BUILT_MODULE): PRIVATE_ANDROID_FS_HDR := $(system_android_filesystem_config)
 $(LOCAL_BUILT_MODULE): $(LOCAL_PATH)/fs_config_generator.py $(TARGET_FS_CONFIG_GEN) $(system_android_filesystem_config)
 	@mkdir -p $(dir $@)
-	$(hide) $< passwd --required-prefix=vendor_ --aid-header=$(PRIVATE_ANDROID_FS_HDR) $(or $(PRIVATE_TARGET_FS_CONFIG_GEN),/dev/null) > $@
+	$(hide) $< passwd --required-prefix=vendor_ --aid-header=$(PRIVATE_ANDROID_FS_HDR) $(allow_legacy_aids) $(or $(PRIVATE_TARGET_FS_CONFIG_GEN),/dev/null) > $@
 
 ##################################
 # Generate the vendor/etc/group text file for the target
@@ -441,7 +447,7 @@ $(LOCAL_BUILT_MODULE): PRIVATE_TARGET_FS_CONFIG_GEN := $(TARGET_FS_CONFIG_GEN)
 $(LOCAL_BUILT_MODULE): PRIVATE_ANDROID_FS_HDR := $(system_android_filesystem_config)
 $(LOCAL_BUILT_MODULE): $(LOCAL_PATH)/fs_config_generator.py $(TARGET_FS_CONFIG_GEN) $(system_android_filesystem_config)
 	@mkdir -p $(dir $@)
-	$(hide) $< group --required-prefix=vendor_ --aid-header=$(PRIVATE_ANDROID_FS_HDR) $(or $(PRIVATE_TARGET_FS_CONFIG_GEN),/dev/null) > $@
+	$(hide) $< group --required-prefix=vendor_ --aid-header=$(PRIVATE_ANDROID_FS_HDR) $(allow_legacy_aids) $(or $(PRIVATE_TARGET_FS_CONFIG_GEN),/dev/null) > $@
 
 system_android_filesystem_config :=
 system_capability_header :=
diff --git a/tools/fs_config/fs_config_generator.py b/tools/fs_config/fs_config_generator.py
index dccff92..6b622fa 100755
--- a/tools/fs_config/fs_config_generator.py
+++ b/tools/fs_config/fs_config_generator.py
@@ -588,7 +588,7 @@ class FSConfigFileParser(object):
     _SECTIONS = [('_handle_aid', ('value', )),
                  ('_handle_path', ('mode', 'user', 'group', 'caps'))]
 
-    def __init__(self, config_files, oem_ranges):
+    def __init__(self, config_files, oem_ranges, warn_only=False):
         """
         Args:
             config_files ([str]): The list of config.fs files to parse.
@@ -596,6 +596,7 @@ class FSConfigFileParser(object):
             oem_ranges ([(),()]): range tuples indicating reserved OEM ranges.
         """
 
+        self._warn_only = warn_only
         self._files = []
         self._dirs = []
         self._aids = []
@@ -710,7 +711,10 @@ class FSConfigFileParser(object):
         if not Utils.in_any_range(int(aid.value, 0), self._oem_ranges):
             emsg = '"value" not in valid range %s, got: %s'
             emsg = emsg % (str(self._oem_ranges), value)
-            sys.exit(error_message(emsg))
+            if self._warn_only:
+                sys.stderr.write(error_message(emsg))
+            else:
+                sys.exit(error_message(emsg))
 
         # use the normalized int value in the dict and detect
         # duplicate definitions of the same value
@@ -994,13 +998,21 @@ class FSConfigGen(BaseGenerator):
 
         opt_group.add_argument('--out_file', required=True, help='Output file')
 
+        opt_group.add_argument(
+            '--allow-legacy-aids',
+            action="store_true",
+            required=False,
+            default=False,
+            help='Allow legacy AIDs that fall outside allowed OEM ranges')
+
     def __call__(self, args):
 
         self._capability_parser = CapabilityHeaderParser(
             args['capability_header'])
         self._base_parser = AIDHeaderParser(args['aid_header'])
         self._oem_parser = FSConfigFileParser(args['fsconfig'],
-                                              self._base_parser.oem_ranges)
+                                              self._base_parser.oem_ranges,
+                                              args['allow_legacy_aids'])
 
         self._partition = args['partition']
         self._all_partitions = args['all_partitions']
@@ -1265,11 +1277,20 @@ class OEMAidGen(BaseGenerator):
             help='An android_filesystem_config.h file'
             'to parse AIDs and OEM Ranges from')
 
+        opt_group.add_argument(
+            '--allow-legacy-aids',
+            action="store_true",
+            required=False,
+            default=False,
+            help='Allow legacy AIDs that fall outside allowed OEM ranges')
+
     def __call__(self, args):
 
         hdr_parser = AIDHeaderParser(args['aid_header'])
 
-        parser = FSConfigFileParser(args['fsconfig'], hdr_parser.oem_ranges)
+        parser = FSConfigFileParser(args['fsconfig'],
+                                    hdr_parser.oem_ranges,
+                                    args['allow_legacy_aids'])
 
         print OEMAidGen._GENERATED
 
@@ -1321,11 +1342,20 @@ class PasswdGen(BaseGenerator):
             required=False,
             help='A prefix that the names are required to contain.')
 
+        opt_group.add_argument(
+            '--allow-legacy-aids',
+            action="store_true",
+            required=False,
+            default=False,
+            help='Allow legacy AIDs that fall outside allowed OEM ranges')
+
     def __call__(self, args):
 
         hdr_parser = AIDHeaderParser(args['aid_header'])
 
-        parser = FSConfigFileParser(args['fsconfig'], hdr_parser.oem_ranges)
+        parser = FSConfigFileParser(args['fsconfig'],
+                                    hdr_parser.oem_ranges,
+                                    args['allow_legacy_aids'])
 
         required_prefix = args['required_prefix']
 
-- 
2.7.4

